src/test/rgw/test_d4n_filter.cc

```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+      std::string instance;
+      put_version_enabled_object(testName, instance, yield); // Upload two versions 
+      instances.push_back(instance);
+    }
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("ZREVRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "-1");
+
+      response< std::vector<std::string> > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ(std::get<0>(resp).value().size(), 2);
+      EXPECT_EQ(std::get<0>(resp).value()[0], instances[1]);
+      EXPECT_EQ(std::get<0>(resp).value()[1], instances[0]);
+    }
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + instances[1], err), true);  
+    std::string oid = instances[1] + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid, err), true);  
+
+    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op_enabled = objEnabled->get_delete_op();
+    objEnabled->set_instance(instances[1]); // Latest version
+    EXPECT_EQ(del_op_enabled->delete_obj(env->dpp, optional_yield{yield}, rgw::sal::FLAG_LOG_OP), 0);
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("ZREVRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "-1");
+
+      response< std::vector<std::string> > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ(std::get<0>(resp).value().size(), 1);
+      EXPECT_EQ(std::get<0>(resp).value()[0], instances[0]);
+    }
+
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(null_yield);
+  }, rethrow);
+
+  io.run_for(std::chrono::seconds(2)); // Allow cleaning cycle to complete
+
+  net::spawn(io, [this, &testName, &bucketName, &instances] (net::yield_context yield) {
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + instances[1], err), false);  
+    std::string oid = instances[1] + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid, err), false);  
+
+    // Make sure deleted instance isn't written to backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), -2);
+    }
+
+    objEnabled->set_instance(instances[0]);
+
+    // First instance should still be written to backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
```

Comment:
where are you verifying the version? You can call next->iterate and see what version is being returned.

Reply:
Yes, I'm setting the instance value to the first instance that wsan't deleted, and using this to get the next object, which should represent that instance in the backend. Since I can load the object and explicitly check if it exists in the line below, I believe that proves that particular instance is available in the backend.

Reply:
If there are two instances and you set an instance (even the second one which is not the latest) and call load_obj_state(), even then you would be able to see the object, so that doesn't check for the correct latest version, hence I asked you to do a simple get and check the latest version.

Reply:
Yes, I added the simple get call for this purpose but should I remove the calls to `next->load_obj_state` then? On line 3224, I did set the instance to the second instance, which has been deleted, and the `next->load_obj_state` call returns -2 to represent this.

Reply:
No what you have done is correct.